### PR TITLE
Oj 1257 access token handler coverage new branch on main

### PR DESCRIPTION
**Pull request by surajkumar**
### What changed

session-service.ts
errors.ts
access-token-handler.test.ts

### Why did it change

session-service does not check if access or auth codes have expired like in the Java versions. This PR adds that and unit tests. 

### Issue tracking

- [OJ-1257](https://govukverify.atlassian.net/browse/OJ-1257)

### Environment variables or secrets

- [X] No environment variables or secrets were added or changed

[OJ-1257]: https://govukverify.atlassian.net/browse/OJ-1257?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/src/types/errors.ts [hunk 1]
```diff
@@ -69,3 +69,19 @@ export class SessionValidationError extends BaseError {
         return `${this.code}: ${this?.details}`;
     }
 }
+
+export class AuthorizationCodeExpiredError extends BaseError {
+    constructor() {
+        super("Authorization code expired");
+        this.statusCode = 403;
+        this.code = 1026;
```
Comment (jkunle) [inline lambdas/src/types/errors.ts:77]: 1027

Comment (surajkumar) [reply lambdas/src/types/errors.ts]: Updated

File: lambdas/src/services/session-service.ts [hunk 2]
```diff
@@ -56,6 +61,14 @@ export class SessionService {
             throw new InvalidAccessTokenError();
         }
 
+        if (sessionItem.Items[0].authorizationCodeExpiryDate < new Date().getDate()) {
+            throw new AuthorizationCodeExpiredError();
+        }
+
+        if (sessionItem.Items[0].accessTokenExpiryDate < new Date().getDate()) {
```
Comment (jkunle) [inline lambdas/src/services/session-service.ts:68]: I think this is out of scope of this ticket, it is the Credential issue lambda that needs to check whether the accesstoken has expired which is further upstream (we haven't converted issuercredential lambda that why the method appeared to be missing, it is not i.e it is the java sessionService). Whereas the Accesstoken lambda needs to know if the authorizationCode created downstream i.e created when /Authorize endpoint was hit) is still valid

File: lambdas/src/types/errors.ts [hunk 3]
```diff
@@ -69,3 +69,19 @@ export class SessionValidationError extends BaseError {
...
+
+export class AuthorizationCodeExpiredError extends BaseError {
+    constructor() {
+        super("Authorization code expired");
+        this.statusCode = 403;
+        this.code = 1027;
+    }
+}
+
+export class AccessCodeExpiredError extends BaseError {
```
Comment (jkunle) [inline lambdas/src/types/errors.ts:81]: Also out of scope due to the above explanation

File: lambdas/tests/unit/handlers/access-token-handler.test.ts [hunk 4]
```diff
@@ -131,6 +139,109 @@ describe("access-token-handler.ts", () => {
                 expect(output.statusCode).toBe(200);
                 expect(metricsSpy).toHaveBeenCalledWith("accesstoken", MetricUnits.Count, 1);
             });
+
+            it("should return http 200 if the authorizationCodeExpiryDate is within date", async () => {
```
Comment (jkunle) [inline lambdas/tests/unit/handlers/access-token-handler.test.ts:143]: This also out of scope

File: lambdas/tests/unit/handlers/access-token-handler.test.ts [hunk 5]
```diff
@@ -131,6 +139,109 @@ describe("access-token-handler.ts", () => {
...
+                        },
+                    } as unknown as APIGatewayProxyEvent,
+                    null,
+                );
+
+                expect(output.statusCode).toBe(200);
+                expect(metricsSpy).toHaveBeenCalledWith("accesstoken", MetricUnits.Count, 1);
+            });
+
+            it("should return http 200 if the accessTokenExpiryDate has not expired", async () => {
```
Comment (jkunle) [inline lambdas/tests/unit/handlers/access-token-handler.test.ts:195]: This is also out of scope

File: lambdas/tests/unit/handlers/access-token-handler.test.ts (hunk 4 above)
Comment (jkunle) [inline lambdas/tests/unit/handlers/access-token-handler.test.ts:143]: We already have a 200 success test, I would rather see a 403 error test instead the whole point of this ticket was to try and increase test coverage for access-token-handler. Also I think you should mock or spy of the date, rather than making a call to getDate, the test could start feel under some circumstances

Comment (surajkumar) [reply lambdas/tests/unit/handlers/access-token-handler.test.ts:143]: I added a 403 test under the "fail paths" alongside a 200 test in the success paths.

**Comment by jkunle** [comment]
I think the focus should more of test coverage increase here, we do have ticket to explore typescript vs java coverage we should avoid this being about that. I details the error by test for in the ticket and they of the class 403 error which there isn't coverage for

File: lambdas/tests/unit/handlers/access-token-handler.test.ts (hunk 5 above)
Comment (surajkumar) [reply lambdas/tests/unit/handlers/access-token-handler.test.ts]: Removed

File: lambdas/src/types/errors.ts (hunk 3 above)
Comment (surajkumar) [reply lambdas/src/types/errors.ts]: Removed

File: lambdas/src/services/session-service.ts (hunk 2 above)
Comment (surajkumar) [reply lambdas/src/services/session-service.ts]: OK - Removed it

File: lambdas/tests/unit/handlers/access-token-handler.test.ts (hunk 4 above)
Comment (jkunle) [reply lambdas/tests/unit/handlers/access-token-handler.test.ts:147]: Please check the ticket details to see what I mean, we can have quick chat if it makes things clearer

Comment (surajkumar) [reply lambdas/tests/unit/handlers/access-token-handler.test.ts:147]: Updated `session-service#getSessionByAuthorizationCode` to check against current epoch time.
As discussed the 403 unit test is fine but I have also added a test to check the 403 when we have multiple session items.

**Review by jkunle — APPROVED** [review]
